### DISCUSSION TITLE
Replacing a custom element with its contents

### DISCUSSION THREAD

**Discussion by revosw** [lit/lit#2169] [Q&A]
If an element `<my-header>` looks like this:
```js
render() {
  html`
    <span>A</span>
    <h1>What does the dot product represent?</h1>
  `
}
```

I want to replace the definition:
```html
<div>
  <my-header></my-header>
</div>
```

With its contents:
```js
<div>
  <span>A</span>
  <h1>What does the dot product represent?</h1>
</div>
```

In the React world, there are React fragments. This lets you return multiple nodes without a parent.
```jsx
return (
  <>
    <span>A</span>
    <h1>What does the dot product represent?</h1>
  </>
)
```
Does lit have something like it?

The closest I can find in lit is to use the connectedCallback like this:
```js
// If there is no shadow DOM
connectedCallback() {
  super.connectedCallback()
  setTimeout(() => this.outerHTML = this.innerHTML, 0)
}

// If there is shadow DOM
connectedCallback() {
  super.connectedCallback()
  setTimeout(() => {
    while(this.childNodes.length) {
      //Some wizardry that places all children of this shadow root up to the same level as the parent
    }
  },0)
}
```
How will such a solution affect `<slot>`?

**Comment by justinfagnani** [comment]
In Lit you would just use a template instead of a component.

The rest of the questions are not Lit-specific, but web components related. In general you shouldn't be replacing your self with content because the host element belongs to the outer element and you shouldn't modify the outer element's DOM. That's mostly a concern with reusable / maximally interoperable components and might not apply to app-specific components.

Why do you want to replace a component with it's content?

**Reply by revosw** [reply]
> Why do you want to replace a component with it's content?

So I'm using tailwind, which I admit after a while goes against everything shadow DOM has to offer. Though I've given up Lit for now, I'll share what I was thinking.

In my design, each heading consists of two components. They're all part of a grid
```html
<span class="question-tab">A</span>
<h1 class="question-text">
  Calculate the magnitude of the position vector $p$ for point $P(4,5,6)$
</h1>
```
![image](https://user-images.githubusercontent.com/19785016/133909537-b86d5eac-1a0e-40ae-9a13-b55fe8aef590.png)
![image](https://user-images.githubusercontent.com/19785016/133909531-02b840ce-d077-42a8-bd09-328625c73ffb.png)

I tried to make a component that combines them.
```html
<question-header-tabbed tab="A">
  Calculate the magnitude of the position vector $p$ for point $P(4,5,6)$
</question-header-tabbed>
```
But since `<question-header-tabbed>` is a parent, the design breaks. So I tried separating the two:
```html
<question-tab>A</question-tab>
<question-text>
  Calculate the magnitude of the position vector $p$ for point $P(4,5,6)$
</question-text>
```
And by now I started questioning whether that was really any beneficial compared to:
```html
<span class="question-tab">A</span>
<h1 class="question-text">
  Calculate the magnitude of the position vector $p$ for point $P(4,5,6)$
</h1>
```
If I hadn't used tailwind, there would of course be the benefit of style encapsulation. But since I'm fighting it, I've found Lit is not the right tool for the job.

**Reply by justinfagnani** [reply]
Ok, well if something similar comes up again you can use `:host {display: contents;}` and expose the two elements in the shadow root as shadow parts and then they should be able to participate in the outer grid.
